registry/repository.go

```diff
@@ -21,176 +20,123 @@ const (
-	return &RepositoryClient{
-		Repository: remote.Repository{
-			Client:    client,
-			Reference: ref,
-			PlainHTTP: plainHTTP,
-		},
+// NewRepository returns a new Repository
+func NewRepository(repo remote.Repository) Repository {
+	return &repositoryClient{
+		Repository: repo,
 	}
 }
 
 // Resolve resolves a reference(tag or digest) to a manifest descriptor
-func (c *RepositoryClient) Resolve(ctx context.Context, reference string) (notation.Descriptor, error) {
-	desc, err := c.Repository.Resolve(ctx, reference)
-	if err != nil {
-		return notation.Descriptor{}, err
-	}
-	return notationDescriptorFromOCI(desc), nil
+func (c *repositoryClient) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	return c.Repository.Resolve(ctx, reference)
 }
 
-// ListSignatureManifests returns all signature manifests given the manifest digest
-func (c *RepositoryClient) ListSignatureManifests(ctx context.Context, manifestDigest digest.Digest) ([]SignatureManifest, error) {
-	var signatureManifests []SignatureManifest
-	if err := c.Repository.Referrers(ctx, ocispec.Descriptor{
-		Digest: manifestDigest,
-	}, ArtifactTypeNotation, func(referrers []artifactspec.Descriptor) error {
-		for _, desc := range referrers {
-			if desc.MediaType != artifactspec.MediaTypeArtifactManifest {
-				continue
-			}
-			artifact, err := c.getArtifactManifest(ctx, desc.Digest)
-			if err != nil {
-				return fmt.Errorf("failed to fetch manifest: %v: %v", desc.Digest, err)
-			}
-			if len(artifact.Blobs) == 0 {
-				continue
-			}
-			signatureManifests = append(signatureManifests, SignatureManifest{
-				Blob:        notationDescriptorFromArtifact(artifact.Blobs[0]),
-				Annotations: artifact.Annotations,
-			})
-		}
-		return nil
-	}); err != nil {
-		return nil, err
-	}
-	return signatureManifests, nil
+// ListSignatures returns signature manifests filtered by fn given the
+// artifact manifest descriptor
+func (c *repositoryClient) ListSignatures(ctx context.Context, desc ocispec.Descriptor, fn func(signatureManifests []ocispec.Descriptor) error) error {
+	return c.Repository.Referrers(ctx, desc, ArtifactTypeNotation, func(referrers []ocispec.Descriptor) error {
+		return fn(referrers)
+	})
 }
 
-// GetBlob downloads the content of the specified digest's Blob
-func (c *RepositoryClient) GetBlob(ctx context.Context, digest digest.Digest) ([]byte, error) {
-	desc, err := c.Repository.Blobs().Resolve(ctx, digest.String())
+// FetchSignatureBlob returns signature envelope blob and descriptor given
+// signature manifest descriptor
+func (c *repositoryClient) FetchSignatureBlob(ctx context.Context, desc ocispec.Descriptor) ([]byte, ocispec.Descriptor, error) {
+	sigManifest, err := c.getSignatureManifest(ctx, desc)
 	if err != nil {
-		return nil, err
+		return nil, ocispec.Descriptor{}, err
 	}
-	if desc.Size > maxBlobSizeLimit {
-		return nil, fmt.Errorf("signature blob too large: %d", desc.Size)
+	if len(sigManifest.Blobs) == 0 {
```

Comment:
The number of blobs in a signature manifest should exactly be **one** as per spec.